### ISSUE TITLE
动态图中如何便捷的进行梯度累加

### ISSUE THREAD

**Issue by HighCWu** [PaddlePaddle/Paddle#25408]
- 版本、环境信息：
    PaddlePaddle版本：1.8.0
- 伪代码
```python
optimizer.clear_gradients()
losses = []
for i in range(8):
  fake_y = model(x)
  loss = layers.mean((fake_y - real_y)**2)
  loss.backward()
  losses.append(loss)
optimizer.minimize(losses)
```
如何实现losses中所有loss的反传梯度累加在一起用于一次性优化？有便捷的实现吗？重新包装一个优化器会有些麻烦。应该不是ReComputeOptimizer那种。
- 对标PyTorch实现
```python
optimizer.zero_grad()
for i in range(8):
  fake_y = model(x)
  loss = layers.mean((fake_y - real_y)**2)
  loss.backward()
optimizer.step()
```
PyTorch的已经实现的过程会简洁很多，因为它会自动计算进行了反向传播的损失

**Comment by phlrain** [comment]
这个 反传梯度累加在一起 是说在loss.backward的时候进行累加么。。这个paddle已经进行了累加了

**Comment by HighCWu** [comment]
> 这个 反传梯度累加在一起 是说在loss.backward的时候进行累加么。。这个paddle已经进行了累加了

@phlrain 不是，多个loss通过backward算好了梯度，然后将这个梯度加起来取平均，优化器的话使用minimize只能输入一个loss，没办法所有的loss一起梯度下降，详情你可以搜索pytorch的梯度累加

**Comment by zhiqiu** [comment]
hi，@HighCWu 不同batch之间的梯度累计，我们会在后面进行支持。

**Comment by HighCWu** [comment]
> hi，@HighCWu 不同batch之间的梯度累计，我们会在后面进行支持。

@zhiqiu 行，那当下版本我应该只能重写优化器了😂

**Comment by Meiyim** [comment]
请问梯度累计这个feature在paddle 2.0中有实现么。

**Comment by HighCWu** [comment]
> 请问梯度累计这个feature在paddle 2.0中有实现么。

2.0可以实现梯度累加了
```python
optimizer.clear_grad()
for i in range(8):
  fake_y = model(x)
  loss = layers.mean((fake_y - real_y)**2) / 8
  loss.backward()
optimizer.step()
```

**Comment by Meiyim** [comment]
@HighCWu 真的有实现么，可否指出是哪一版PR合入的功能。
我的测试代码如下：
```
import paddle as p
f=p.nn.Linear(3,3)
x=p.randn([3,3])
y=f(x).mean()
for _ in range(3):
        y.backward(retain_graph=True)
        print(f.weight.gradient())
```
从输出看并没有累计(最新编译develop分支)：
```
W1126 12:33:07.353596     1 device_context.cc:338] Please NOTE: device: 0, CUDA Capability: 70, Driver API Version: 10.2, Runtime API Version: 10.2
W1126 12:33:07.371047     1 device_context.cc:346] device: 0, cuDNN Version: 7.6.
[[ 0.10692619  0.10692619  0.10692619]
 [ 0.2788641   0.2788641   0.2788641 ]
 [-0.32339698 -0.32339698 -0.32339698]]
[[ 0.10692619  0.10692619  0.10692619]
 [ 0.2788641   0.2788641   0.2788641 ]
 [-0.32339698 -0.32339698 -0.32339698]]
[[ 0.10692619  0.10692619  0.10692619]
 [ 0.2788641   0.2788641   0.2788641 ]
 [-0.32339698 -0.32339698 -0.32339698]]
```

**Comment by HighCWu** [comment]
可能我理解错了，我以为现在用step()后会自动累积了，看样子没有，可能只是做了个表面，我之后测试一下😂

**Comment by zhwesky2010** [comment]
@HighCWu 最新的RC1版本已经支持了不同batch间的梯度累加，麻烦重装试试

**Comment by Meiyim** [comment]
把PR在这里贴一下吧@zhouwei25

**Comment by zhwesky2010** [comment]
@Meiyim @HighCWu  https://github.com/PaddlePaddle/Paddle/pull/28429 实现了
```
# 不同batch间梯度累加
x = paddle.to_tensor(5., stop_gradient=False)
for i in range(10):
    y = paddle.pow(x, 4.0)
    y.backward()
    print("{}: {}".format(i, x.grad))

0: [500.]
1: [1000.]
2: [1500.]
3: [2000.]
4: [2500.]
5: [3000.]
6: [3500.]
7: [4000.]
8: [4500.]
9: [5000.]

x.clear_gradient()
print("{}".format(x.grad))

0.
```

**Comment by HighCWu** [comment]
在cuda 10.1下使用paddle 2.0.0rc1测试成功。

过程如下：

- 下载whl文件

```bash
wget https://paddle-wheel.bj.bcebos.com/2.0.0-rc1-gpu-cuda10.1-cudnn7-mkl_gcc8.2%2Fpaddlepaddle_gpu-2.0.0rc1.post101-cp37-cp37m-linux_x86_64.whl -O paddlepaddle_gpu-2.0.0rc1.post101-cp37-cp37m-linux_x86_64.whl
```

- 安装paddlepaddle gpu

```bash
pip install paddlepaddle_gpu-2.0.0rc1.post101-cp37-cp37m-linux_x86_64.whl
```

- 单一graph多次运行测试梯度累积

输入：

```python
import paddle as p
p.seed(233)
f=p.nn.Linear(3,3)
x=p.randn([3,3])
y=f(x).mean()
for i in range(3):
    y.backward(retain_graph=True)
    print(f.weight.grad)
```

输出：
```
[[-0.11581856 -0.11581856 -0.11581856]
 [ 0.08599838  0.08599838  0.08599838]
 [ 0.29122534  0.29122534  0.29122534]]
[[-0.23163712 -0.23163712 -0.23163712]
 [ 0.17199676  0.17199676  0.17199676]
 [ 0.5824507   0.5824507   0.5824507 ]]
[[-0.34745568 -0.34745568 -0.34745568]
 [ 0.25799513  0.25799513  0.25799513]
 [ 0.87367606  0.87367606  0.87367606]]
```

- 多次构造graph的梯度累积测试

```python
import paddle as p
p.seed(233)
f=p.nn.Linear(3,3)
x=p.randn([3,3])
for i in range(3):
    y=f(x).mean()
    y.backward()
    print(f.weight.grad)
```

输出：
```
[[-0.11581856 -0.11581856 -0.11581856]
 [ 0.08599838  0.08599838  0.08599838]
 [ 0.29122534  0.29122534  0.29122534]]
[[-0.23163712 -0.23163712 -0.23163712]
 [ 0.17199676  0.17199676  0.17199676]
 [ 0.5824507   0.5824507   0.5824507 ]]
[[-0.34745568 -0.34745568 -0.34745568]
 [ 0.25799513  0.25799513  0.25799513]
 [ 0.87367606  0.87367606  0.87367606]]
```

验证完成，@Meiyim 没有疑问我将关闭这个issue

**Comment by arcral** [comment]
> @HighCWu 最新的RC1版本已经支持了不同batch间的梯度累加，麻烦重装试试

batch norm操作也能梯度累加吗？？？
